File: core/src/main/java/org/evergreen_ils/accountAccess/bookbags/BookBagActivity.java
/*
 * Copyright (C) 2012 Evergreen Open-ILS
 * @author Daniel-Octavian Rizea
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * or the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be usefull,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 * 
 */
package org.evergreen_ils.accountAccess.bookbags;

import java.util.ArrayList;
import java.util.List;

import android.app.AlertDialog;
import android.view.WindowManager;
import org.evergreen_ils.R;
import org.evergreen_ils.accountAccess.AccountAccess;
import org.evergreen_ils.accountAccess.SessionNotFoundException;
import org.evergreen_ils.system.Analytics;
import org.evergreen_ils.system.Log;
import org.evergreen_ils.system.Utils;
import org.evergreen_ils.utils.ui.BaseActivity;
import org.evergreen_ils.utils.ui.ProgressDialogSupport;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

public class BookBagActivity extends BaseActivity {

    private final static String TAG = BookBagActivity.class.getSimpleName();

    private AccountAccess accountAccess = null;

    private ListView lv;

    private BookBagsArrayAdapter listAdapter = null;

    private ArrayList<BookBag> bookBags = null;

    private ProgressDialogSupport progress;

    private EditText bookbag_name;

    private Button create_bookbag;

    private Runnable getBookbagsRunnable;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (mRestarting) return;

        setContentView(R.layout.activity_bookbags);

        // prevent soft keyboard from popping up when the activity starts
        this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);

        accountAccess = AccountAccess.getInstance();
        progress = new ProgressDialogSupport();

        bookbag_name = findViewById(R.id.bookbag_create_name);
        create_bookbag = findViewById(R.id.bookbag_create_button);
        create_bookbag.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                createBookbag(bookbag_name.getText().toString());
            }
        });

        lv = findViewById(R.id.bookbag_list);
        bookBags = new ArrayList<BookBag>();
        listAdapter = new BookBagsArrayAdapter(this, R.layout.bookbag_list_item, bookBags);
        lv.setAdapter(listAdapter);
        lv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Analytics.logEvent("Lists: Tap List");
                BookBag item = (BookBag) lv.getItemAtPosition(position);
                Intent intent = new Intent(BookBagActivity.this, BookBagDetailsActivity.class);
                intent.putExtra("bookBag", item);
                startActivityForResult(intent, 0);
            }
        });

        initGetBookbagsRunnable();

        new Thread(getBookbagsRunnable).start();
    }

    @Override
    protected void onDestroy() {
        if (progress != null) progress.dismiss();
        super.onDestroy();
    }

    private void initGetBookbagsRunnable() {
        getBookbagsRunnable = new Runnable() {
            @Override
            public void run() {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        if (!isFinishing()) progress.show(BookBagActivity.this, getString(R.string.msg_retrieving_lists));
                    }
                });

                try {
                    accountAccess.retrieveBookbags();
                } catch (SessionNotFoundException e) {
                    try {
                        if (accountAccess.reauthenticate(BookBagActivity.this))
                            accountAccess.retrieveBookbags();
                    } catch (Exception e2) {
                        Log.d(TAG, "caught", e2);
                    }
                }
                bookBags = accountAccess.getBookbags();

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listAdapter.clear();
                        for (BookBag bookBag : bookBags)
                            listAdapter.add(bookBag);

                        listAdapter.notifyDataSetChanged();

                        progress.dismiss();
                    }
                });
            }
        };
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (resultCode) {
        case BookBagDetailsActivity.RESULT_CODE_UPDATE:
            new Thread(getBookbagsRunnable).start();
            break;
        }
    }

    private void createBookbag(final String name) {
        if (name.length() < 2) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.msg_list_name_too_short_title)
                    .setMessage(R.string.msg_list_name_too_short)
                    .setPositiveButton(android.R.string.ok, null);
            builder.create().show();
            return;
        }
        Analytics.logEvent("Lists: Create List");

        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    accountAccess.createBookbag(name);
                } catch (SessionNotFoundException e) {
                    try {
                        if (accountAccess.reauthenticate(BookBagActivity.this))
                            accountAccess.createBookbag(name);
                    } catch (Exception eauth) {
                        Log.d(TAG, "caught", eauth);
                    }
                }

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        progress.dismiss();
                    }
                });

                new Thread(getBookbagsRunnable).start();
            }
        });

        progress.show(this, getString(R.string.msg_creating_list));
        thread.start();
    }

    class BookBagsArrayAdapter extends ArrayAdapter<BookBag> {
        private List<BookBag> records;

        public BookBagsArrayAdapter(Context context, int textViewResourceId, List<BookBag> objects) {
            super(context, textViewResourceId, objects);
            this.records = objects;
        }

        public int getCount() {
            return this.records.size();
        }

        public BookBag getItem(int index) {
            return this.records.get(index);
        }

        public View getView(int position, View convertView, ViewGroup parent) {
            View row = convertView;

            // Get item
            final BookBag record = getItem(position);

            // if it is the right type of view
            if (row == null) {
                LayoutInflater inflater = (LayoutInflater) this.getContext()
                        .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                row = inflater.inflate(R.layout.bookbag_list_item, parent, false);
            }

            TextView nameText = (TextView) row.findViewById(R.id.bookbag_name);
            nameText.setText(Utils.safeString(record.name));
            TextView descText = (TextView) row.findViewById(R.id.bookbag_description);
            descText.setText(Utils.safeString(record.description));
            TextView itemsText = (TextView) row.findViewById(R.id.bookbag_items);
            itemsText.setText(getResources().getQuantityString(R.plurals.number_of_items,
                    record.items.size(), record.items.size()));

            return row;
        }
    }
}

File: core/src/main/java/org/evergreen_ils/accountAccess/bookbags/BookBagActivity.kt
/*
 * Copyright (C) 2012 Evergreen Open-ILS
 * @author Daniel-Octavian Rizea
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * or the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be usefull,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 * 
 */
package org.evergreen_ils.accountAccess.bookbags

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.joinAll
import org.evergreen_ils.R
import org.evergreen_ils.accountAccess.AccountAccess
import org.evergreen_ils.accountAccess.SessionNotFoundException
import org.evergreen_ils.android.Analytics
import org.evergreen_ils.android.App
import org.evergreen_ils.android.Log
import org.evergreen_ils.data.BookBag
import org.evergreen_ils.data.Result
import org.evergreen_ils.net.Gateway
import org.evergreen_ils.utils.StringUtils
import org.evergreen_ils.utils.ui.BaseActivity
import org.evergreen_ils.utils.ui.ProgressDialogSupport
import org.evergreen_ils.utils.ui.showAlert
import java.util.*

class BookBagActivity : BaseActivity() {
    private var accountAccess: AccountAccess? = null
    private var lv: ListView? = null
    private var listAdapter: BookBagsArrayAdapter? = null
    private var bookBags: ArrayList<BookBag>? = null
    private var progress: ProgressDialogSupport? = null
    private var bookbag_name: EditText? = null
    private var create_bookbag: Button? = null
    private var getBookbagsRunnable: Runnable? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (isRestarting) return

        setContentView(R.layout.activity_bookbags)

        // prevent soft keyboard from popping up when the activity starts
        this.window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)

        accountAccess = AccountAccess.getInstance()
        progress = ProgressDialogSupport()
        bookbag_name = findViewById(R.id.bookbag_create_name)
        create_bookbag = findViewById(R.id.bookbag_create_button)
        create_bookbag?.setOnClickListener(View.OnClickListener {
            createBookbag(bookbag_name?.getText().toString())
        })
        lv = findViewById(R.id.bookbag_list)
        bookBags = ArrayList()
        listAdapter = BookBagsArrayAdapter(this, R.layout.bookbag_list_item, bookBags!!)
        lv?.setAdapter(listAdapter)
        lv?.setOnItemClickListener(AdapterView.OnItemClickListener { parent, view, position, id ->
            Analytics.logEvent("Lists: Tap List")
            val item = lv?.getItemAtPosition(position) as BookBag
            val intent = Intent(this@BookBagActivity, BookBagDetailsActivity::class.java)
            intent.putExtra("bookBag", item)
            startActivityForResult(intent, 0)
        })
        initGetBookbagsRunnable()
        Thread(getBookbagsRunnable).start()
    }

    override fun onDestroy() {
        progress?.dismiss()
        super.onDestroy()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        Log.d(TAG, object{}.javaClass.enclosingMethod?.name)
        fetchData()
    }

    private fun fetchData() {
        async {
            try {
                Log.d(TAG, "[kcxxx] fetchData ...")
                val start = System.currentTimeMillis()
                var jobs = mutableListOf<Job>()
                progress?.show(this@BookBagActivity, getString(R.string.msg_retrieving_lists))

                // fetch bookbags
                val result = Gateway.actor.fetchUserBookBags(App.getAccount())
                when (result) {
                    is Result.Success -> {
                        bookBags = BookBag.makeArray(result.data)
                    }
                    is Result.Error -> {
                        showAlert(result.exception)
                        return@async
                    }
                }

                jobs.joinAll()
                //updateList()
                Log.logElapsedTime(TAG, start, "[kcxxx] fetchData ... done")
            } catch (ex: Exception) {
                Log.d(TAG, "[kcxxx] fetchData ... caught", ex)
                showAlert(ex)
            } finally {
                progress?.dismiss()
            }
        }
    }

    private fun initGetBookbagsRunnable() {
        /*
        getBookbagsRunnable = Runnable {
            runOnUiThread { if (!isFinishing) progress!!.show(this@BookBagActivity, getString(R.string.msg_retrieving_lists)) }
            try {
                accountAccess!!.retrieveBookbags()
            } catch (e: SessionNotFoundException) {
                try {
                    if (accountAccess!!.reauthenticate(this@BookBagActivity)) accountAccess!!.retrieveBookbags()
                } catch (e2: Exception) {
                    Log.d(TAG, "caught", e2)
                }
            }
            bookBags = accountAccess!!.bookbags
            runOnUiThread {
                listAdapter!!.clear()
                for (bookBag in bookBags) listAdapter!!.add(bookBag)
                listAdapter!!.notifyDataSetChanged()
                progress!!.dismiss()
            }
        }
        */
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (resultCode) {
            BookBagDetailsActivity.RESULT_CODE_UPDATE -> Thread(getBookbagsRunnable).start()
        }
    }

    private fun createBookbag(name: String) {
        if (name.length < 2) {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.msg_list_name_too_short_title)
                    .setMessage(R.string.msg_list_name_too_short)
                    .setPositiveButton(android.R.string.ok, null)
            builder.create().show()
            return
        }
        Analytics.logEvent("Lists: Create List")
        val thread = Thread(Runnable {
            try {
                accountAccess!!.createBookbag(name)
            } catch (e: SessionNotFoundException) {
                try {
                    if (accountAccess!!.reauthenticate(this@BookBagActivity)) accountAccess!!.createBookbag(name)
                } catch (eauth: Exception) {
                    Log.d(TAG, "caught", eauth)
                }
            }
            runOnUiThread { progress!!.dismiss() }
            Thread(getBookbagsRunnable).start()
        })
        progress!!.show(this, getString(R.string.msg_creating_list))
        thread.start()
    }

    internal inner class BookBagsArrayAdapter(context: Context, private val resourceId: Int, private val items: List<BookBag>) : ArrayAdapter<BookBag>(context, resourceId, items) {

        override fun getCount(): Int {
            return items.size
        }

        override fun getItem(index: Int): BookBag {
            return items[index]
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
            var row = when(convertView) {
                null -> {
                    val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
                    inflater.inflate(resourceId, parent, false)
                }
                else -> {
                    convertView
                }
            }

            val nameText = row.findViewById<View>(R.id.bookbag_name) as TextView
            val descText = row.findViewById<View>(R.id.bookbag_description) as TextView
            val itemsText = row.findViewById<View>(R.id.bookbag_items) as TextView

            val record = getItem(position)
            nameText.text = StringUtils.safeString(record.name)
            descText.text = StringUtils.safeString(record.description)
            itemsText.text = resources.getQuantityString(R.plurals.number_of_items,
                    record.items!!.size, record.items!!.size)

            return row
        }

    }

    companion object {
        private val TAG = BookBagActivity::class.java.simpleName
    }
}